CDAP-18010: Introduce ArtifactInspector interface

wyzhang opened the pull request:
Why:
We can have a default implementation and remote execution implementation
which allows this to run in isolation (e.g. in task worker k8s pod),
thus better security and potentially more scalable.

CuriousVini commented on cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/artifact/ArtifactInspector.java:
```diff
@@ -0,0 +1,49 @@
+   * @throws InvalidArtifactException if the artifact is invalid. For example, if the application main class is not
+   * actually an Application.
+   */
+  ArtifactClassesWithMetadata inspectArtifact(Id.Artifact artifactId, File artifactFile,
```
should this be location in case of remote `ArtifactInspector`?

wyzhang replied:
Currently the only place where inspection takes place is right after artifact uploaded, before it is added to artifact store and copied to location. So inspection today only happens on a temp file.  

Expect the same for remote execution, during which file gets uploaded to remote server as a local tmp file for inspection.

wyzhang committed:
CDAP-18010: Introduce ArtifactInspector interface

Why:
We can have a default implementation and remote execution implementation
which allows this to run in isolation (e.g. in task worker k8s pod),
thus better security and potentially more scalable.
